```{r}
.libPaths(c('/usr/local/lib/R/site-library', '/usr/local/lib/R/library', '/tmp/RtmpTPz6Xo/downloaded_packages'))
#wget https://cf.10xgenomics.com/samples/cell-exp/3.0.0/neuron_10k_v3/neuron_10k_v3_filtered_feature_bc_matrix.h5
```

```{r "setup"}
set.seed(61)
options(future.globals.maxSize = 4000 * 1024^5)
library(RColorBrewer)
library(Seurat)
library(patchwork)
library(ggplot2)
library(dplyr)
library(hdf5r)
library(stringr)
library(biomaRt)
library(kableExtra)
library(knitr)
library(pdftools)
library(viridis)
library(openxlsx)
library(SeuratDisk)
library(SeuratData)
library(ComplexHeatmap)
library(plotly)
```

```{r "import data"}
#wget https://cf.10xgenomics.com/samples/cell-exp/3.0.2/5k_pbmc_v3/5k_pbmc_v3_filtered_feature_bc_matrix.h5
#wget https://cf.10xgenomics.com/samples/cell-exp/3.0.0/pbmc_10k_v3/pbmc_10k_v3_filtered_feature_bc_matrix.h5
#wget https://cf.10xgenomics.com/samples/cell-exp/3.0.0/pbmc_1k_v3/pbmc_1k_v3_filtered_feature_bc_matrix.h5

pbmc.1k <- Read10X_h5('/home/rstudio/data/pbmc_1k_v3_filtered_feature_bc_matrix.h5')
pbmc.10k <- Read10X_h5('/home/rstudio/data/pbmc_10k_v3_filtered_feature_bc_matrix.h5')
```

```{r "create seurat object"}
pbmc.1k <- CreateSeuratObject(counts = pbmc.1k, project = 'pbmc.1k')
pbmc.10k <- CreateSeuratObject(counts = pbmc.10k, project = 'pbmc.10k')
```

#https://satijalab.org/seurat/articles/merge_vignette.html
#merge() merges the raw count matrices of two Seurat objects and creates a new Seurat object with the resulting combined raw count matrix. To easily tell which original object any particular cell came from, you can set the add.cell.ids parameter with an c(x, y) vector, which will prepend the given identifier to the beginning of each cell name. The original project ID will remain stored in object meta data under orig.ident
```{r "merge"}
all_data <- merge(x = pbmc.1k, y = c(pbmc.10k), project = 'pbmc')
```

#the format of the mt sequences will vary depending on which organism/genome is used...
```{r "add mt percent data"}
all_data[["percent.mt"]] <- PercentageFeatureSet(all_data, pattern = "^MT-")
```

```{r}
#Before we plot, set the order of the object idents to whatever order we'd like
Idents(all_data) <- 'orig.ident'
levels(all_data) <- c("pbmc.1k", "pbmc.10k")

```

#https://www.google.com/url?sa=t&rct=j&q=&esrc=s&source=web&cd=&ved=2ahUKEwi9243gqoX3AhVvneAKHVD5B60QFnoECAUQAQ&url=https%3A%2F%2Fwww.bioinformatics.babraham.ac.uk%2Ftraining%2F10XRNASeq%2FIntroduction.pdf&usg=AOvVaw0d7gdoHTH4l3zVJgH5ikXD
#nFeature_RNA is the number of genes 
#nCount_RNA is the number of UMIs (unique molecules -- like counts) 
```{r  "QC plots"}


VlnPlot(all_data, features = "nFeature_RNA")
VlnPlot(all_data, features = "nCount_RNA")
VlnPlot(all_data, features="percent.mt")

FeatureScatter(all_data, feature1 = "nCount_RNA", feature2 = "percent.mt")
FeatureScatter(all_data, feature1 = "nFeature_RNA", feature2 = "percent.mt")
FeatureScatter(all_data, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
#we can see groups of cells with:
#high mt count, lower UMI counts (nCount) and lower unique genes (nFeature)


```

#pick your filtering based on your qc plots, mt filtering based on your understanding of the system or pubs

```{r "Data filtering"}
all_data_sub <- subset(all_data, subset = nFeature_RNA > 500 & nFeature_RNA < 600 & percent.mt < 25)
```


#use SCTransform to normalize -- use as an alternative to the NormalizeData, FindVariableFeatures, ScaleData workflow
```{r "SCTransform"}
all_data_split <- SplitObject(all_data_sub, split.by = 'orig.ident')

all_data_list <- lapply(all_data_split, function(x) {
    x <- SCTransform(x,verbose=FALSE)}) # if you run this you might get several iteration limit reached issues



```

```{r "Integration"}
all_data_list.features <- SelectIntegrationFeatures(object.list=all_data_list)

all_data_list <- PrepSCTIntegration(object.list=all_data_list, anchor.features=all_data_list.features, verbose=FALSE)

all_data_list <- lapply(X = all_data_list, FUN = RunPCA, features = all_data_list.features)

ifnb.list <- lapply(X = ifnb.list, FUN = RunPCA, features = features)

all_data.int.anchors <- FindIntegrationAnchors(all_data_list, normalization.method="SCT", anchor.features=all_data_list.features, verbose=FALSE)

all_data.combined <- IntegrateData(anchorset=all_data.int.anchors, normalization.method="SCT", verbose=FALSE)

DefaultAssay(all_data.combined) <- "integrated"

```

```{r "PCA"}
all_data.combined <- RunPCA(all_data.combined)

ElbowPlot(all_data.combined, ndims = 50)
#20 dims

VizDimLoadings(all_data.combined, dims = 1:5, reduction = "pca", nfeatures = 5)

DimPlot(all_data.combined, reduction = 'pca', group.by = 'orig.ident')
```

```{r "clustering"}
all_data.combined <- FindNeighbors(all_data.combined, dims = 1:30)
all_data.combined <- FindClusters(all_data.combined)
all_data.combined <- RunUMAP(all_data.combined, dims = 1:30)
all_data.combined <- RunTSNE(all_data.combined, tsne.method = "FIt-SNE", seed.use=61)
```

```{r "look at clusters"}
DimPlot(all_data.combined, reduction='tsne', split.by = 'orig.ident')
DimPlot(all_data.combined, reduction='umap', split.by = 'orig.ident')

```
```{r "look at a few summary tables to of cells per orig.ident and seurat cluster}
table(all_data.combined@meta.data$orig.ident)
table(all_data.combined@meta.data$seurat_clusters)
```
```{r "FindAllMarkers to find DE genes across clusters or per orig.ident}
Idents(all_data.combined) <- 'seurat_clusters'
DefaultAssay(all_data.combined) <- 'RNA'
cluster_markers <- FindAllMarkers(all_data.combined, only.pos = TRUE, assay = 'RNA')

Idents(all_data.combined) <- 'orig.ident'
ident_markers <- FindAllMarkers(all_data.combined, only.pos = TRUE, assay = 'RNA')
```

```{r "Look at some featureplots"}

FeaturePlot(all_data.combined, features = 'Msp300', reduction = 'tsne', split.by = 'orig.ident', order = T, keep.scale = 'all') & 
   scale_colour_gradientn(colours = rev(brewer.pal(n = 11, name = "Spectral"))) & 
   theme(legend.position = "right")

FeaturePlot(all_data.combined, features = 'Msp300', reduction = 'tsne', split.by = 'orig.ident', order = T, keep.scale = 'all') & 
   scale_colour_gradientn(limits = c(0,300),colours = rev(brewer.pal(n = 11, name = "Spectral"))) & 
   theme(legend.position = "right")

```
#still need to add bits about how to add annotations from cell atlas data

