---
title: "Introduction"
output:
  html_document:
    df_print: paged
  html_notebook: default
  pdf_document: default
bibliography: bibliography.bib
---

# scRNAseq in R Workshop

This repository contains R markdown files for the analysis of scRNAseq data (mostly in Seurat). It was developed by the [Computational Biology Core](http://cbc.brown.edu/) and is based mainly on the Seurat vignettes https://satijalab.org/seurat and GitHub repository https://github.com/satijalab/seurat.

### Learning Objectives

- Understanding the difference between bulk and scRNAseq.
- Familiarity with 10x Genomics gene expression and multimodal data library preparation.
- Familiarity with cellranger outputs and importing them into R.
- Deep knowledge of Seurat object structure and data storage.
- Understanding of metrics for QCing and filtering data.
- Merging and integrating multiple datasets, the difference between them, and when they are appropriate to use.
- Understanding how to classify cell types using Seurat TransferAnchors and the distinction between data transfer and data integration.
- Understanding of how and why we normalize data using SCTransform.
- Data visualization functions in Seurat and their default behaviors.
- Understanding of parallelization options in Seurat and other R packages.
- Importing CITE-seq data in Seurat.
- How to run CLR transformation of CITE-seq data and the distinction between normalizing over features or cells.
- A brief introduction to some up-and-coming methods to control for Type I error rate in estimated clusters
- Understanding the distinction between over representation analysis and gene set enrichment and their appropriateness for scRNAseq analysis.

### Getting started

** Running the notebook in Open on Demand at Brown **
** Running the notebook on a local machine **

